### DISCUSSION TITLE
exploring-the-dotnet-8-preview-avoiding-flaky-tests-with-timeprovider-and-itimer/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#393] [Announcements]
# exploring-the-dotnet-8-preview-avoiding-flaky-tests-with-timeprovider-and-itimer/

In this post I look at the new TimeProvider and ITimer types added in .NET 8, describe how to use them, and show some of the testing problems they solve

https://andrewlock.net/exploring-the-dotnet-8-preview-avoiding-flaky-tests-with-timeprovider-and-itimer/

**Comment by Stolpe** [comment]
How can we specify which timezone is to be considered "local"? I haven't had the opportunity to play around with this, but _my_ current (IE pre .Net 8) solution to this is yet another pattern, where I inject an options type specifying the time zone I want to treat as my local. I prefer to not depend on the settings of the host machine, if that is possible. Is there some way (in DI initiation?) to say "this is what I consider my local time zone"?

Current solution, which uses the TimeZoneConverter NuGet, as not all hosts contain good time zone info:
```
public interface ITimeService : IUtcTimeService
{
    DateTime GetNow();
}

public interface IUtcTimeService
{
    DateTime GetUtcNow();
}

public class TimeOptions
{
    public string TimeZone { get; set; }
}

public class TimeService : ITimeService
{
    private readonly IUtcTimeService _utcTimeService;
    private readonly TimeOptions _options;

    public TimeService(
        IUtcTimeService utcTimeService,
        IOptions<TimeOptions> options)
    {
        _utcTimeService = utcTimeService;
        _options = options.Value;
    }

    public DateTime GetNow() =>
        TimeZoneInfo.ConvertTimeFromUtc(
            GetUtcNow(),
            LocalTimeZone);

    public DateTime GetUtcNow() =>
        _utcTimeService.GetUtcNow();

    private TimeZoneInfo LocalTimeZone =>
        TZConvert.GetTimeZoneInfo(
            _options.TimeZone);
}

public class UtcTimeService : IUtcTimeService
{
    public DateTime GetUtcNow() =>
        DateTime.UtcNow;
}
```

**Reply by quicoli** [reply]
@Stolpe have you seen this?

https://github.com/dariusz-wozniak/FakeLocalTimeZone

**Reply by andrewlock** [reply]
The `TimeProvider` abstraction supports this - you can control what is configured the local time zone. [The announcement blog post showed an example of this](https://devblogs.microsoft.com/dotnet/announcing-dotnet-8-preview-4/#usage-examples), it would just be a case of creating a wrapper around the `SystemTimeProvider` and wiring it up in DI

**Reply by Stolpe** [reply]
@quicoli: Thanks for your tip. This package does not take away the dependence of host settings, though. This may handled by a different team than the developers, so being able to set the time zone via settings takes away that dependency.

@andrewlock: That's neat. Thanks for showing me this example :)

**Comment by agehrke** [comment]
Yeah, the `TimeProvider` and the functionality to create timers paired with the `FakeTimeProvider` is great for *triggering* timers in tests. I often find myself doing async work in timer callbacks and thus can't rely on that work being done when `Advance()` method returns as in your example. Instead I need to observe the completion of that work in another way and then make my assertions. One might argue that testing the functionality performed in the timer callback should be more explicitly, eg by invoking some method encapsulating that functionality, but that's a topic for another day.
 
May I add that a TimeProvider can also be passed to `CancellationTokenSource` which can help when testing timeouts implemented via a `CancellationToken`.

**Comment by mh-logiten** [comment]
Thanks for the post. There is not a massive amount of information out there currently surrounding testing timers/delays with the FakeTimeProvider - it's been instrumental in improving my tests.

The race condition with "Advance" & "Task.Delay" cost me some time, there is another alternative to using a timer directly. It is not a catch all for every circumstance but it worked well enough for my purposes.

You can create a derivative of FakeTimeProvider and override the CreateTimer method. In the following example the time is automatically advanced to match the delay and the delay is recorded allowing your tests to count the number of delays and the length of each:

``` csharp
public class AutoDelayTimeProvider : FakeTimeProvider
{
    public IList<TimeSpan> Delays { get; } = new List<TimeSpan>();
    
    public override ITimer CreateTimer(
        TimerCallback callback,
        object? state,
        TimeSpan dueTime,
        TimeSpan period)
    {
        var timer = base.CreateTimer(callback, state, dueTime, period);
        Delays.Add(dueTime);
        Advance(dueTime);
        return timer;
    }
}
```

**Reply by ivanxpetrov** [reply]
Great point. Another variation of this is to have Delays be a BlockingCollection (maybe ConcurrentQueue, not sure) and actually don't auto Advance, but have AdvanceToNextTimer 

```csharp

public void AdvanceToNextTimer() {
     var delay = _queue.Take();
     Advance(delay ); // Advance
}
```

**Comment by aksoftware98** [comment]
Amazing article, thank you so much for the great details and the examples.

**Comment by mjnamjkaze** [comment]
Good article, but
 public bool IsExpired() => _expiresDate > DateTime.UtcNow;
should be
 public bool IsExpired() => _expiresDate < DateTime.UtcNow;
is it right ?
